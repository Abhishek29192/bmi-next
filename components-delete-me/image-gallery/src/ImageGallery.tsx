import React, { useState } from "react";
import classnames from "classnames";
import { Media } from "@bmi-digital/components";
import { Typography } from "@bmi-digital/components";
import { Truncate } from "@bmi-digital/components";
import DesktopThumbnails from "./_DesktopThumbnails";
import MobileThumbnails from "./_MobileThumbnails";
import { Image } from "./types";
import styles from "./ImageGallery.module.scss";

type Props = {
  images: readonly Image[];
  imageSize?: "cover" | "contain";
  thumbnailComponent?: React.ComponentType<any>; // TODO
  layout?: "default" | "short";
  className?: string;
};

const renderMedia = (
  { media }: Image,
  imageSize: Props["imageSize"],
  layout?: Props["layout"]
) => {
  const className = classnames(
    styles["main-image-wrapper"],
    imageSize !== "contain" && styles[`main-image-wrapper--${imageSize}`],
    styles[`main-image-wrapper--${layout}`]
  );

  if (media) {
    return (
      <Media size={imageSize} className={className}>
        {media}
      </Media>
    );
  }

  return null;
};

const ImageGallery = ({
  images,
  imageSize = "contain",
  thumbnailComponent,
  layout = "default",
  className
}: Props) => {
  if (!images.length) {
    return null;
  }

  const [activeImageIndex, setActiveImageIndex] = useState<number>(0);
  const isTouchDevice =
    typeof document !== `undefined` &&
    "ontouchstart" in document.documentElement;

  const Thumbnails = isTouchDevice ? MobileThumbnails : DesktopThumbnails;
  // eslint-disable-next-line security/detect-object-injection
  const caption = images[activeImageIndex].caption;

  return (
    <div className={classnames(styles["ImageGallery"], className)}>
      <div className={styles["image-wrapper"]}>
        {/* eslint-disable-next-line security/detect-object-injection */}
        {renderMedia(images[activeImageIndex], imageSize, layout)}
        {caption ? (
          <div className={styles["caption"]}>
            <Typography
              variant="h6"
              component="p"
              className={styles["caption-text"]}
            >
              <Truncate lines="2">{caption}</Truncate>
            </Typography>
          </div>
        ) : null}
      </div>
      {images.length > 1 && (
        <Thumbnails
          images={images}
          component={thumbnailComponent}
          activeImageIndex={activeImageIndex}
          onThumbnailClick={setActiveImageIndex}
        />
      )}
    </div>
  );
};

export default ImageGallery;
