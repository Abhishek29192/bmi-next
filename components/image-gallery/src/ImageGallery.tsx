import React, { useState } from "react";
import classnames from "classnames";
import Media from "@bmi/media";
import Typography from "@bmi/typography";
import Truncate from "@bmi/truncate";
import DesktopThumbnails from "./_DesktopThumbnails";
import MobileThumbnails from "./_MobileThumbnails";
import { Image } from "./types";
import styles from "./ImageGallery.module.scss";

type Props = {
  images: Image[];
  imageSize?: "cover" | "contain";
  thumbnailComponent?: React.ComponentType<any>; // TODO
  layout?: "default" | "short";
};

const renderMedia = (
  { mainSource, altText, media }: Image,
  imageSize: Props["imageSize"],
  layout?: Props["layout"]
) => {
  const className = classnames(
    styles["main-image-wrapper"],
    {
      [styles[`main-image-wrapper--${imageSize}`]]: imageSize !== "contain"
    },
    styles[`main-image-wrapper--${layout}`]
  );

  if (mainSource && altText) {
    // TODO: Deprecate this case.
    return (
      <div
        className={className}
        style={{
          backgroundImage: `url(${mainSource})`
        }}
      >
        <span className={styles["accessibility-label"]}>{altText}</span>
      </div>
    );
  }

  if (media) {
    return (
      <Media size={imageSize} className={className}>
        {media}
      </Media>
    );
  }

  return null;
};

const ImageGallery = ({
  images,
  imageSize = "contain",
  thumbnailComponent,
  layout = "default"
}: Props) => {
  if (!images.length) {
    return null;
  }

  const [activeImageIndex, setActiveImageIndex] = useState<number>(0);
  let isTouchDevice =
    typeof document !== `undefined` &&
    "ontouchstart" in document.documentElement;

  const Thumbnails = isTouchDevice ? MobileThumbnails : DesktopThumbnails;

  return (
    <div className={styles["ImageGallery"]}>
      <div className={styles["image-wrapper"]}>
        {renderMedia(images[activeImageIndex], imageSize, layout)}
        {images[activeImageIndex].caption ? (
          <div className={styles["caption"]}>
            <Typography
              variant="h6"
              component="p"
              className={styles["caption-text"]}
            >
              <Truncate lines="2">{images[activeImageIndex].caption}</Truncate>
            </Typography>
          </div>
        ) : null}
      </div>
      {images.length > 1 && (
        <Thumbnails
          images={images}
          component={thumbnailComponent}
          activeImageIndex={activeImageIndex}
          onThumbnailClick={setActiveImageIndex}
        />
      )}
    </div>
  );
};

export default ImageGallery;
