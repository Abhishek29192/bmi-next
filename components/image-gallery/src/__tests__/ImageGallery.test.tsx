import React from "react";
import { render } from "@testing-library/react";
import ImageGallery from "../";
import mockImage from "./images/demo-tiles.jpg";

describe("ImageGallery component", () => {
  it("renders correctly with default layout", () => {
    const images = [
      {
        media: <img src={mockImage} alt="Lorem ipsum" />,
        thumbnail: mockImage
      },
      {
        media: <img src={mockImage} alt="Lorem ipsum" />,
        thumbnail: mockImage
      },
      {
        media: <img src={mockImage} alt="Lorem ipsum" />,
        thumbnail: mockImage
      }
    ];

    const { container } = render(<ImageGallery images={images} />);
    expect(container).toMatchSnapshot();
  });
  it("renders null when no images are passed", () => {
    const { container } = render(<ImageGallery images={[]} />);
    expect(container).toMatchSnapshot();
  });
  it("renders without a thumbnail when one image is passed", () => {
    const { container } = render(
      <ImageGallery
        images={[
          {
            media: <img src={mockImage} alt="Lorem ipsum" />,
            altText: "Single image"
          }
        ]}
      />
    );
    expect(container).toMatchSnapshot();
  });
  it("renders correctly with short layout", () => {
    const images = [
      {
        mainSource: mockImage,
        altText: "Demo Tiles",
        thumbnail: mockImage
      },
      {
        mainSource: mockImage,
        altText: "Demo Tiles Black"
      },
      {
        mainSource: mockImage,
        altText: "Demo house"
      }
    ];

    const { container } = render(
      <ImageGallery images={images} layout="short" />
    );
    expect(container).toMatchSnapshot();
  });
  it("renders empty image gallery when no media and main source are supplied", () => {
    const { container } = render(<ImageGallery images={[{}]} />);
    expect(container).toMatchSnapshot();
  });
  it("renders caption of first image when caption is present", () => {
    const images = [
      {
        altText: "Demo Tiles",
        thumbnail: mockImage,
        caption:
          "test caption : renders this caption from first image as this is default active index!!"
      },
      {
        thumbnail: mockImage,
        altText: "Demo Tiles Black"
      },
      {
        thumbnail: mockImage,
        altText: "Demo house"
      }
    ];

    const { container } = render(
      <ImageGallery images={images} layout="short" />
    );
    expect(container).toMatchSnapshot();
  });
  it("renders with main-image-wrapper--cover style if imageSize is not contain", () => {
    const images = [
      {
        media: <img src={mockImage} alt="Lorem ipsum" />,
        thumbnail: mockImage
      }
    ];

    const { container } = render(
      <ImageGallery images={images} imageSize="cover" />
    );

    expect(container.querySelector(".main-image-wrapper--cover")).toBeTruthy();
  });
  it("renders MobileThumbnails if isTouchDevice is true", () => {
    Object.defineProperty(document.documentElement, "ontouchstart", jest.fn());
    const images = [
      {
        media: <img src={mockImage} alt="Lorem ipsum" />,
        thumbnail: mockImage
      },
      {
        media: <img src={mockImage} alt="Lorem ipsum" />,
        thumbnail: mockImage
      }
    ];
    const { container } = render(<ImageGallery images={images} />);
    expect(container.querySelector(".thumbnails--touch")).toBeTruthy();
  });
});
