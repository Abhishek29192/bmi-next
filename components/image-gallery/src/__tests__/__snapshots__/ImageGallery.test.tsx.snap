// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageGallery component renders caption of first image when caption is present 1`] = `
<div>
  <div
    class="ImageGallery"
  >
    <div
      class="image-wrapper"
    >
      <div
        class="caption"
      >
        <p
          class="MuiTypography-root caption-text Typography MuiTypography-h6"
        >
          <span
            class="LinesEllipsis  Truncate"
          >
            test caption : renders this caption from first image as this is default active index!!
            <wbr />
          </span>
        </p>
      </div>
    </div>
    <div
      class="thumbnails"
    >
      <button
        class="MuiButtonBase-root thumb-scroller thumb-scroller--left ThumbScrollerButton ThumbScrollerButton--left"
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root thumb-scroller thumb-scroller--right thumb-scroller--hidden ThumbScrollerButton ThumbScrollerButton--right"
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <div
        class="scroller"
        style="margin-right: 0%;"
      >
        <button
          class="MuiButtonBase-root thumbnail Thumbnail Thumbnail--selected Thumbnail--large"
          style="background-image: url(path-to-image);"
          tabindex="0"
          type="button"
        >
          <span
            class="accessibility-text"
          >
            Demo Tiles
          </span>
        </button>
        <button
          class="MuiButtonBase-root thumbnail Thumbnail Thumbnail--large"
          style="background-image: url(path-to-image);"
          tabindex="0"
          type="button"
        >
          <span
            class="accessibility-text"
          >
            Demo Tiles Black
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root thumbnail Thumbnail Thumbnail--large"
          style="background-image: url(path-to-image);"
          tabindex="0"
          type="button"
        >
          <span
            class="accessibility-text"
          >
            Demo house
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ImageGallery component renders correctly with default layout 1`] = `
<div>
  <div
    class="ImageGallery"
  >
    <div
      class="image-wrapper"
    >
      <img
        alt="Lorem ipsum"
        class="main-image-wrapper main-image-wrapper--default Media Media--contain"
        src="path-to-image"
      />
    </div>
    <div
      class="thumbnails"
    >
      <button
        class="MuiButtonBase-root thumb-scroller thumb-scroller--left ThumbScrollerButton ThumbScrollerButton--left"
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root thumb-scroller thumb-scroller--right thumb-scroller--hidden ThumbScrollerButton ThumbScrollerButton--right"
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <div
        class="scroller"
        style="margin-right: 0%;"
      >
        <button
          class="MuiButtonBase-root thumbnail Thumbnail Thumbnail--selected Thumbnail--large"
          style="background-image: url(path-to-image);"
          tabindex="0"
          type="button"
        >
          <img
            alt="Lorem ipsum"
            class="Media"
            src="path-to-image"
          />
        </button>
        <button
          class="MuiButtonBase-root thumbnail Thumbnail Thumbnail--large"
          style="background-image: url(path-to-image);"
          tabindex="0"
          type="button"
        >
          <img
            alt="Lorem ipsum"
            class="Media"
            src="path-to-image"
          />
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root thumbnail Thumbnail Thumbnail--large"
          style="background-image: url(path-to-image);"
          tabindex="0"
          type="button"
        >
          <img
            alt="Lorem ipsum"
            class="Media"
            src="path-to-image"
          />
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ImageGallery component renders correctly with short layout 1`] = `
<div>
  <div
    class="ImageGallery"
  >
    <div
      class="image-wrapper"
    />
    <div
      class="thumbnails"
    >
      <button
        class="MuiButtonBase-root thumb-scroller thumb-scroller--left ThumbScrollerButton ThumbScrollerButton--left"
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root thumb-scroller thumb-scroller--right thumb-scroller--hidden ThumbScrollerButton ThumbScrollerButton--right"
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <div
        class="scroller"
        style="margin-right: 0%;"
      >
        <button
          class="MuiButtonBase-root thumbnail Thumbnail Thumbnail--selected Thumbnail--large"
          style="background-image: url(path-to-image);"
          tabindex="0"
          type="button"
        >
          <span
            class="accessibility-text"
          >
            Demo Tiles
          </span>
        </button>
        <button
          class="MuiButtonBase-root thumbnail Thumbnail Thumbnail--large"
          tabindex="0"
          type="button"
        >
          <span
            class="accessibility-text"
          >
            Demo Tiles Black
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root thumbnail Thumbnail Thumbnail--large"
          tabindex="0"
          type="button"
        >
          <span
            class="accessibility-text"
          >
            Demo house
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ImageGallery component renders empty image gallery when no media and main source are supplied 1`] = `
<div>
  <div
    class="ImageGallery"
  >
    <div
      class="image-wrapper"
    />
  </div>
</div>
`;

exports[`ImageGallery component renders null when no images are passed 1`] = `<div />`;

exports[`ImageGallery component renders without a thumbnail when one image is passed 1`] = `
<div>
  <div
    class="ImageGallery"
  >
    <div
      class="image-wrapper"
    >
      <img
        alt="Lorem ipsum"
        class="main-image-wrapper main-image-wrapper--default Media Media--contain"
        src="path-to-image"
      />
    </div>
  </div>
</div>
`;
